src/engine.ml

```diff
@@ -602,23 +601,29 @@ let incoming_control_client config rng session channel now op data =
           match Config.(find Ifconfig config) with
           | Some _ ->
               let ip_config = ip_from_config config in
+              let cipher = Config.get Cipher config
+              and hmac_algorithm = Config.get Auth config in
+              let keys = kdf session cipher hmac_algorithm key tls_data in
               let channel_st = Established keys in
               Ok (Some ip_config, config, { channel with channel_st }, tls_out)
           | None ->
               (* now we send a PUSH_REQUEST\0 and see what happens *)
               push_request tls' >>| fun (tls'', out) ->
-              let channel_st = Push_request_sent (tls'', keys) in
+              let channel_st = Push_request_sent (tls'', key, tls_data) in
```

Comment:
Here the `tls_data` seems to be the one sent from the server.

Reply:
As far as I understand, we delay this `tls_data` to the `Push_request_sent` which will do the `kdf` on it as we do in the `Some _` branch. So what is the issue here?

Reply:
If I remember correctly I was confused with the `tls_data` in `Push_request_sent _` -- I think in some cases it is our `tls_data` while on other cases it's theirs.

Reply:
I think the trouble come from the `incoming_control_server` and the `incoming_control_client`. But if we keep only the execution path of the `incoming_control_client`, the only way to construct a `Pull_request_sent` is done via what the server give to us. And again, it just matter of a delay to compute `kdf` _after_ `PUSH_REQUEST\0` if it's needed.

Reply:
I hope e36d0ba398312bc2d074e02843fab7ee0acf2f1c clarifies this, there's own_key_material and peer_key_material now :)

and in `kex_server`, indeed the wrong bits were attached to Push_request_sent.